src/katgpucbf/xbgpu/beamform.py

```diff
@@ -0,0 +1,151 @@
+
+"""Implement the calculations for beamforming."""
+
+from importlib import resources
+from typing import Sequence
+
+import numpy as np
+from katsdpsigproc import accel
+from katsdpsigproc.abc import AbstractCommandQueue, AbstractContext
+
+from .. import COMPLEX, N_POLS
+
+
+class BeamformTemplate:
+    """Template for beamforming.
+
+    Parameters
+    ----------
+    context
+        The GPU context that we'll operate in.
+    pols
+        One entry per single-polarisation output beam. Each entry is either
+        0 or 1, to indicate which input polarisation to use in the beam.
+    """
+
+    def __init__(self, context: AbstractContext, beam_pols: Sequence[int]) -> None:
+        # TODO: tune these. And maybe adapt to input shape?
+        self.block_time = 128
+        self.block_channels = 1
+        self.beam_pols = beam_pols
+        with resources.as_file(resources.files(__package__)) as resource_dir:
+            program = accel.build(
+                context,
+                "kernels/beamform.mako",
+                {
+                    "block_time": self.block_time,
+                    "block_channels": self.block_channels,
+                    "beam_pols": self.beam_pols,
+                },
+                extra_dirs=[str(resource_dir)],
+            )
+        self.kernel = program.get_kernel("beamform")
+
+    def instantiate(
+        self,
+        command_queue: AbstractCommandQueue,
+        frames: int,
+        antennas: int,
+        channels: int,
+        times: int,
+    ) -> "Beamform":
+        """Generate a :class:`Beamform` object based on the template."""
+        return Beamform(self, command_queue, frames, antennas, channels, times)
+
+
+class Beamform(accel.Operation):
+    r"""Operation for beamforming.
+
+    .. rubric:: Slots
+
+    **in** : frames × antennas × channels × times × N_POLS × COMPLEX, int8
+        Complex (Gaussian integer) input channelised voltages
+    **out** : frames × beams × channels × times × COMPLEX, int8
+        Complex (Gaussian integer) output channelised voltages
+    **weights** : antennas × beams, complex64
+        Complex scale factor to apply to each antenna for each beam
+    **delays** : antennas × beams, float32
+        Delay used to compute channel-dependent phase rotation. The
+        rotation applied is :math:`e^{\pi cd}` where :math:`c` is
+        the channel number and :math:`d` is the delay value. Note
+        that this will not apply any rotation to the first channel
+        in the data; any such rotation needs to be baked into **weights**.
+
+    Parameters
+    ----------
+    template
+        The template for the operation
+    command_queue
+        The command queue on which to enqueue the work
+    frames
+        Number of instances of the problem (coarse time dimension)
+    antennas
+        Number of antennas
+    channels
+        Number of frequency channels
+    times
```

Comment:
I'm half-convinced. On the other hand, this code should make sense on its own independently of how results are transmitted, and heaps are specific to SPEAD. I'll think about it.

Reply:
I'm going with `n_spectra_per_frame`. It ends up being the same as `n_spectra_per_heap` (because heaps and frames are the same size on the time axis) but avoids referring to heaps.

Reply:
Love a good compromise. Please and thank you.